### PR TITLE
Use sfserviceinstance in dashboard template

### PR DESCRIPTION
**Pull request by amitmalav**

### CODE REVIEW

File: api-controllers/ServiceBrokerApiController.js [hunk 1]
```diff
@@ -215,7 +216,8 @@ class ServiceBrokerApiController extends FabrikBaseController {
           return eventmesh.apiServerClient.getOSBResourceOperationStatus(lastOperationState);
         }
       })
-      .then(done);
+      .then(() => _.get(context, 'plan.manager.settings.dashboard_url_template') !== undefined ? eventmesh.apiServerClient.waitTillInstanceIsScheduled(req.params.instance_id) : {})
```
Comment (subhankarc) [inline api-controllers/ServiceBrokerApiController.js]: update does not need it. You'll have `clusterId` set already.

File: broker/config/settings.yml [hunk 2]
```diff
@@ -782,7 +782,7 @@ defaults: &defaults
           update_predecessors:
           - 'bc158c9a-7934-401e-94ab-057082a5073f'
           template: <%= base64_template('templates/blueprint-manifest.yml.ejs') %>
-          dashboard_url_template: <%= new Buffer('http://service-fabrik-broker.bosh-lite.com/manage/dashboards/director/instances/${instance_id}?planName=${plan.name}&serviceId=${plan.service.id}').toString('base64') %>
+          dashboard_url_template: <%= new Buffer('${spec.clusterId == 1 ? \'http://service-fabrik-broker.bosh-lite.com/manage/dashboards/director/instances/\'+metadata.name+\'?planId=\'+spec.planId+\'&serviceId=\'+spec.serviceId : \'\'}').toString('base64') %>
```
Comment (subhankarc) [inline broker/config/settings.yml]: I am little hesitant about two things here.

1. We are not passing context anymore. Are we missing out any attribute ? I guess no.
2. ABAP colleagues use this feature. They have to adapt when changing to this version as the keys are changed. 

Should we keep context and add `sfserviceinstance`'s `clusterId` as it's sub attribute ?

File: common/constants.js [hunk 3]
```diff
@@ -52,6 +52,7 @@ module.exports = Object.freeze({
     UPDATE_EVENT: 'update_instance'
   },
   EVENTMESH_POLLER_DELAY: 200,
+  CLUSTER_SCHEDULER_DELAY: 50,
```
Comment (subhankarc) [inline common/constants.js]: Isn't it too low? let us keep 200 ms

File: data-access-layer/eventmesh/ApiServerClient.js [hunk 4]
```diff
@@ -206,6 +206,26 @@ class ApiServerClient {
         };
       });
   }
+  /**
+     * @description Waits till clusterId is set for sfserviceinstance
+     * @param {string} resourceId - id of resource
+     */
+  waitTillInstanceIsScheduled(resourceId) {
```
Comment (vivekzhere) [inline data-access-layer/eventmesh/ApiServerClient.js]: Should we add a timeout for the wait also

File: data-access-layer/eventmesh/ApiServerClient.js [hunk 5]
```diff
@@ -206,6 +206,26 @@ class ApiServerClient {
...
+    assert.ok(resourceId, 'Argument \'resourceId\' is required to get scheduled cluster for instance');
+    logger.debug(`Waiting for scheduler to set ClusterId on ${resourceId}`);
+    return Promise.delay(CONST.CLUSTER_SCHEDULER_DELAY)
+      .then(() => this.getResource({
+        resourceGroup: CONST.APISERVER.RESOURCE_GROUPS.INTEROPERATOR,
+        resourceType: CONST.APISERVER.RESOURCE_TYPES.INTEROPERATOR_SERVICEINSTANCES,
+        resourceId: resourceId,
+        namespaceId: this.getNamespaceId(resourceId)
+      }))
+      .then(sfserviceinstance => _.get(sfserviceinstance, 'spec.clusterId') ? sfserviceinstance : this.waitTillInstanceIsScheduled(resourceId))
```
Comment (subhankarc) [inline data-access-layer/eventmesh/ApiServerClient.js]: We should break after 50/55 seconds

File: broker/config/settings.yml (hunk 2 above)
Comment (amitmalav) [reply broker/config/settings.yml]: Changed it such that status quo is maintained. Now you can access `sfserviceinstance` values inside template by using `instance` objejct
Example: `instance.spec.clusterId`

File: api-controllers/ServiceBrokerApiController.js (hunk 1 above)
Comment (amitmalav) [reply api-controllers/ServiceBrokerApiController.js]: Done it considering in future if operators want to change dashboard on plan update etc.

File: common/constants.js (hunk 3 above)
Comment (amitmalav) [reply common/constants.js]: Delay will be too much if we keep it 200 I feel. `100`?

Comment (subhankarc) [reply common/constants.js]: okay

**Comment by vinaybheri** [comment]
Validation on k8s cluster: n failed.

**Comment by vinaybheri** [comment]
Validation on k8s cluster: n-2 is succeeded.

**Comment by vinaybheri** [comment]
Validation on k8s cluster: n-1 is succeeded.

**Pull request merged**